ray-operator/controllers/ray/expectations/scale_expectation_test.go

```diff
@@ -0,0 +1,87 @@
+package expectations
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientFake "sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+func TestRayClusterExpectations(t *testing.T) {
+	setupTest()
+	ctx := context.TODO()
+	fakeClient := clientFake.NewClientBuilder().WithRuntimeObjects().Build()
+	exp := NewRayClusterScaleExpectation(fakeClient)
```

Comment:
I don’t quite understand why our design needs to use a client and actually create an object. Take `scale_expectations_test.go` as an example: you can see that the test covers very simple things without involving any k8s operations. I believe our usage is almost identical, so we shouldn’t need to use the k8s client either.

https://github.com/openkruise/kruise/blob/c426ed9b1ea3565b1a18e2adba628a2d9b956d91/pkg/util/expectations/scale_expectations_test.go

Reply:
The fundamental reason for adding a client was that we used `GenerateName` when creating a Pod, rather than knowing the PodName beforehand like with CloneSet. The `ObserveScale()` method relies on the PodName specified in `ExpectScale()`. 
https://github.com/openkruise/kruise/blob/c426ed9b1ea3565b1a18e2adba628a2d9b956d91/pkg/controller/cloneset/sync/cloneset_scale.go#L205
https://github.com/openkruise/kruise/blob/c426ed9b1ea3565b1a18e2adba628a2d9b956d91/pkg/controller/cloneset/cloneset_event_handler.go#L75

In the design of CloneSet's Expectations, the call to `ScaleExpectations()` must occur before Create or Delete a Pod _(if` ScaleExpectations()` is called after Create or Delete, it may result in `ObserveScale()` being invoked in `podEventHandler` before `ScaleExpectations()`)_. Therefore, with `GenerateName`, we cannot obtain the PodName before Create. This requires modifying the Expectation design to allow for calling `ScaleExpectations()` after creating the Pod.

Similar issues can be observed in the design of ReplicaSet, which utilize `UIDTrackingControllerExpectations` to track pods, but currently do not address UID tracking during pod creation. Introducing expectations similar to those in ReplicaSets would complicate the code.
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L582-L587
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L632-L638

The addition of a client leads to simpler code implementation, as we no longer need to handle complex `ObserveScale` logic in the `podEventHandler`. We simply check if the pod in the informer meets the Expectation set during the last reconcile before proceeding with reconciliation and releasing the satisfied Expectation.
The concern here might be whether the client becomes a performance bottleneck. In the code, if there was a Create or Delete of a Pod in the last reconcile, the next reconcile will call `client.Get() `to check if the created or deleted Pod meets the conditions, and delete the Expectation upon success. Under low watch latency, the number of client calls is roughly in line with the number of successful pod Create/Delete. In cases of high watch latency, repeated checks may occur more frequently. However, our current practical validation shows that it does not cause a significant performance bottleneck.

Reply:
Thank you for your response. `GenerateName` is indeed an issue. The reason I asked is that I’m concerned your approach may not solve the original problem. The original problem was that we only read from the informer cache, which is often stale, leading to many unnecessary recreations. But now, in expectations, you’re still using the `controller-runtime` client to read. If I’m not mistaken, the controller runtime client reads from the informer cache, so wouldn’t this approach fail to address the original issue?

Whether it’s Kubernetes or Kruise's expectations implementation, if you look closely, you’ll see that the underlying storage is essentially a thread-safe map. It doesn’t interact with the k8s apiserver or read from the informer cache.

After thinking about it, I don’t think we need to worry about `GenerateName`. Similar to how a replicaset only manages the number of pods without tracking individual pods, our worker group is quite similar to a replicaset, so we could also just manage the quantity. We’d only need to change the key. The replicaset controller’s key is namespace + replicaset name, which ensures uniqueness within the cluster. We could change our key to something like namespace + raycluster name + worker group name. The rest should work similarly to a replicaset, where we only care about the final number of pods, not individual pods. What do you think of this approach?

Reply:
I believe the original issue can be resolved. In non-test code, both Expectation and controller use the same client, which reads from the informer for Pods. The Expectation logic is that during reconciliation, if a Pod is created, it expects the informer to have that Pod. This ensures that when the controller checks the informer for Pod counts during scaling operations, it always gets a consistent number with etcd.

If the number of pods is used as the expectation, we need to incorporate `UIDTrackingControllerExpectations`, `ControllerExpectations`, and `ControlleeExpectations` in the replicaset. From Kubernetes community experience, simply managing counts can lead to unexpected issues. We still need to include UIDTracking logic. This is because multiple updates to pod `deleteTimestamp` can result in the `PodEventHandler` receiving multiple change events for the same Pod. This will result in `DeletionObserved()` being called multiple times. Therefore, the replicaset tracks the pod name rather than the replicas when deleting a Pod.
https://github.com/kubernetes/kubernetes/issues/22270#issuecomment-192134872
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L443

I'm not against introducing the expectations used by replicaset. However, we need to implement `PodEventHandler` additionally.

Reply:
Thanks for the explanation. Makes sense to me. Let me review your code now.